### PR TITLE
[ONNX] Use the stable APIs in onnxscript and sync the latest logic

### PR DESCRIPTION
**Pull request by justinchuby**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #134782

Use the stable apis from onnxscript: https://github.com/microsoft/onnxscript/issues/1827
Sync with torch-onnx at https://github.com/justinchuby/torch-onnx/compare/v0.1.12...v0.1.15.

### CODE REVIEW

File: torch/onnx/_internal/exporter/_registration.py [hunk 1]
```diff
@@ -162,33 +159,19 @@ def from_torchlib(
...
+                    domain,
+                    name,
+                )
+                onnx_decomposition = OnnxDecompMeta(
+                    onnx_function=overload_func,
+                    fx_target=target,
+                    is_custom=False,
+                    is_complex=meta.is_complex,
+                )
+                registry._register(target, onnx_decomposition)
```
Comment (titaiwangms) [inline torch/onnx/_internal/exporter/_registration.py:174]: Is there still registry api for custom ops?

Comment (titaiwangms) [reply torch/onnx/_internal/exporter/_registration.py:174]: It just occurs to me that we might need to refactor ONNX tutorial on PyTorch/tutorial afer the migration.

File: torch/onnx/_internal/exporter/_compat.py [hunk 2]
```diff
@@ -206,20 +189,22 @@ def export_compat(
...
 
-    if f is not None and should_convert_version:
-        assert opset_version is not None
-        if verbose is not False:
-            print(
-                f"[torch.onnx] Converting the ONNX file to opset version {opset_version}..."
-            )
-        _convert_version(f, opset_version)
+    # Converter opset version and optimize
+    onnx_program.model = onnxscript_apis.convert_version(
```
Comment (titaiwangms) [inline torch/onnx/_internal/exporter/_compat.py:197]: Check if the model is already in the desired version or not?

File: torch/onnx/_internal/exporter/_compat.py [hunk 3]
```diff
@@ -206,20 +189,22 @@ def export_compat(
...
+        onnx_program.model, opset_version
+    )
+    onnx_program.model = onnxscript_apis.optimize(onnx_program.model)
+
+    if f is not None:
```
Comment (titaiwangms) [inline torch/onnx/_internal/exporter/_compat.py:202]: f is None is deleted in main api

File: torch/onnx/_internal/exporter/_core.py [hunk 4]
```diff
@@ -768,7 +766,7 @@ def _exported_program_to_onnx_program(
             },
         ),
         ir_version=9,
```
Comment (titaiwangms) [inline torch/onnx/_internal/exporter/_core.py:768]: Should we hide this for version control?

File: torch/onnx/_internal/exporter/_registration.py (hunk 1 above)
Comment (justinchuby) [reply torch/onnx/_internal/exporter/_registration.py:174]: I agree. One thing is how to do want to expose the registration API which I don’t have a good idea yet

File: torch/onnx/_internal/exporter/_compat.py (hunk 2 above)
Comment (justinchuby) [reply torch/onnx/_internal/exporter/_compat.py:197]: I have delegated this to the api itself. So the api is going to do nothing if the model version is desired

File: torch/onnx/_internal/exporter/_compat.py (hunk 3 above)
Comment (justinchuby) [reply torch/onnx/_internal/exporter/_compat.py:202]: I should should change that

File: torch/onnx/_internal/exporter/_core.py (hunk 4 above)
Comment (justinchuby) [reply torch/onnx/_internal/exporter/_core.py:768]: Can you elaborate?

Comment (titaiwangms) [reply torch/onnx/_internal/exporter/_core.py:768]: I am thinking ir version should go with onnxscript version, as it's coded in onnxscript. It seems less informative when we hardcode it here. Would we have to compare ir version in debugging?

Comment (justinchuby) [reply torch/onnx/_internal/exporter/_core.py:768]: IR version is mainly used for enabling the latest dtypes. Hard coding to 9 seems ok for this pytorch version as it has all the dtypes we need.

**Review by titaiwangms — APPROVED** [review]

File: torch/onnx/_internal/exporter/_onnx_program.py [hunk 5]
```diff
@@ -5,26 +5,24 @@
...
 
 
-onnx = _lazy_import.onnx
-ir = _lazy_import.onnxscript_ir
-
-
 if TYPE_CHECKING:
     import onnxruntime as ort
 
+_LARGE_MODEL_THRESHOLD = 512 * 1024 * 1024  # 512MB
```
Comment (justinchuby) [inline torch/onnx/_internal/exporter/_onnx_program.py:24]: Change to 1.5gb

File: torch/onnx/_internal/exporter/_compat.py (hunk 3 above)
Comment (justinchuby) [reply torch/onnx/_internal/exporter/_compat.py:201]: f can be None right? That just means the output is a nonnxprogram

Comment (justinchuby) [reply torch/onnx/_internal/exporter/_compat.py:201]: That's the same signature in `torch.onnx.__init__`

**Pull request closed without merging**